adapters/richaudience/richaudience.go

```diff
@@ -0,0 +1,303 @@
+
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	richaudienceRequest.Imp = resImps
+
+	setSite(request, &richaudienceRequest)
+
+	err = setDevice(request, &richaudienceRequest)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	err = setUser(request, &richaudienceRequest, requestInfo)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	req, err := json.Marshal(richaudienceRequest)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	requestData := &adapters.RequestData{
+		Method:  "POST",
+		Uri:     a.endpoint,
+		Body:    req,
+		Headers: raiHeaders,
+	}
+
+	return []*adapters.RequestData{requestData}, raiErrors
+}
+
+func (a *adapter) MakeBids(request *openrtb2.BidRequest, requestData *adapters.RequestData, responseData *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+
+	if responseData.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if responseData.StatusCode == http.StatusBadRequest {
+		err := &errortypes.BadInput{
+			Message: "Unexpected status code: 400. Bad request from publisher. Run with request.debug = 1 for more info.",
+		}
+		return nil, []error{err}
+	}
+
+	if responseData.StatusCode != http.StatusOK {
+		err := &errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info.", responseData.StatusCode),
+		}
+		return nil, []error{err}
+	}
+
+	var bidResp openrtb2.BidResponse
+	if err := json.Unmarshal(responseData.Body, &bidResp); err != nil {
+
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: err.Error(),
+		}}
+	}
+
+	var response openrtb2.BidResponse
+	if err := json.Unmarshal(responseData.Body, &response); err != nil {
+		return nil, []error{err}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(len(request.Imp))
+	bidResponse.Currency = response.Cur
+	for _, seatBid := range response.SeatBid {
+		for i := range seatBid.Bid {
+			b := &adapters.TypedBid{
+				Bid:     &seatBid.Bid[i],
+				BidType: openrtb_ext.BidTypeBanner,
+			}
+			bidResponse.Bids = append(bidResponse.Bids, b)
+		}
+	}
+
+	return bidResponse, nil
+}
+
+func setHeaders(raiHeaders *http.Header) {
+	raiHeaders.Set("Content-Type", "application/json;charset=utf-8")
+	raiHeaders.Set("Accept", "application/json")
+	raiHeaders.Add("X-Openrtb-Version", "2.5")
+}
+
+func setImp(request *openrtb2.BidRequest, richaudienceRequest *richaudienceRequest) (resImps []openrtb2.Imp, err error) {
+	for _, imp := range request.Imp {
+		var secure = int8(0)
+		raiExt, errImp := parseImpExt(&imp)
+
+		if errImp != nil {
+			return nil, errImp
+		}
+
+		if raiExt != nil {
+			if raiExt.Pid != "" {
+				imp.TagID = raiExt.Pid
+			}
+
+			if raiExt.Test {
+				richaudienceRequest.Test = int8(1)
+			}
+
+			if raiExt.BidFloorCur == "" {
+				imp.BidFloorCur = "USD"
+			} else {
+				imp.BidFloorCur = raiExt.BidFloorCur
+			}
+		}
+
+		if request.Site != nil && request.Site.Page != "" {
+			pageURL, err := url.Parse(request.Site.Page)
```

Comment:
Small optimization: both `setImp` and `setSite` call `pageURL, err := url.Parse(request.Site.Page)` and particularly inside `setImp`, where the call is inside a `for` loop, it may be worth avoiding it. I leave this to you but maybe we could switch the order in which `setImp` and `setSite` get executed inside `MakeRequests`, have `setSite` return a value to use inside `setImp`. In a nutshell:

Switch so `setSite` gets called before `setImp` inside `MakeRequests`:
```
61   func (a *adapter) MakeRequests(request *openrtb2.BidRequest, requestInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
62       richaudienceRequest := richaudienceRequest{}
63   
64       raiHeaders := http.Header{}
65   
66       raiErrors := make([]error, 0, len(request.Imp))
67   
68       setHeaders(&raiHeaders)
69   
70       richaudienceRequest.ID = request.ID
71   
   +     isUrlSecure := setSite(request, &richaudienceRequest)
   +
72 -     resImps, err := setImp(request, &richaudienceRequest)
   +     resImps, err := setImp(request, &richaudienceRequest, isUrlSecure)
73   
74       if err != nil {
75           return nil, []error{err}
76       }
77   
78       richaudienceRequest.Imp = resImps
79   
80 -     setSite(request, &richaudienceRequest)
81   
adapters/richaudience/richaudience.go
```

Make `setSite` be the only place where `pageURL, err := url.Parse(request.Site.Page)` gets called so it only gets executed once and return `isUrlSecure`.
```
209 - func setSite(request *openrtb2.BidRequest, richaudienceRequest *richaudienceRequest) {
    + func setSite(request *openrtb2.BidRequest, richaudienceRequest *richaudienceRequest) (isUrlSecure bool) {
210       if request.Site != nil {
211           if request.Site.Page != "" {
212 -             if request.Site.Domain == "" {
213 -                 pageURL, err := url.Parse(request.Site.Page)
214 -                 if err == nil {
215 -                     richaudienceRequest.Site.Domain = pageURL.Host
216 -                 }
217 -             } else {
212 -             if request.Site.Domain != "" {
218 -                 richaudienceRequest.Site.Domain = request.Site.Domain
219 -             }
220  
    +             richaudienceRequest.Site.Domain = request.Site.Domain
    +             pageURL, err := url.Parse(request.Site.Page)
    +             if err == nil {
    +                 if request.Site.Domain == "" {
    +                     richaudienceRequest.Site.Domain = pageURL.Host
    +                 } 
    +                 isUrlSecure = pageURL.Scheme == "https"
    +             }
221               richaudienceRequest.Site.Page = request.Site.Page
222           }
223       }
224   }
adapters/richaudience/richaudience.go
```

Therefore, we can avoid the `url.Parse(request.Site.Page)` call inside `setImp`, we simply pass the `isUrlSecure` boolean value:
```
161 - func setImp(request *openrtb2.BidRequest, richaudienceRequest *richaudienceRequest) (resImps []openrtb2.Imp, err error) {
    + func setImp(request *openrtb2.BidRequest, richaudienceRequest *richaudienceRequest, isUrlSecure bool) (resImps []openrtb2.Imp, err error) {
162       for _, imp := range request.Imp {
163           var secure = int8(0)
164           raiExt, errImp := parseImpExt(&imp)
165
166 +-- 19 lines: if errImp != nil {-----------------------------
185
186 -         if request.Site != nil && request.Site.Page != "" {
187 -             pageURL, err := url.Parse(request.Site.Page)
188 -             if err == nil && pageURL.Scheme == "https" {
189 -                 secure = int8(1)
190 -             }
191 -         }
    +         if isUrlSecure {
    +             secure = int8(1)
    +         }
192           imp.Secure = &secure
193  
194           if imp.Banner.W == nil && imp.Banner.H == nil {
195               if len(imp.Banner.Format) == 0 {
196                   err = &errortypes.BadInput{
197                       Message: "Format Object not found",
198                   }
199                   return nil, err
200               }
201           }
202  
203           resImps = append(resImps, imp)
204  
205       }
206       return resImps, nil
207   }
adapters/richaudience/richaudience.go
```

This is just an idea, please feel free to modify as you see fit.

Reply:
in next commit i have solved